### ISSUE TITLE
wordpress compatibility

### ISSUE THREAD

**Issue by blindpet** [comotion/VSF#17]
I have been testing VSF with Varnish 4.1 and Wordpress and get some false positives. In your cmd.vcl the cat (third curl) is greedy and captures category. I have modified it to be

```
if (req.url ~ "(=|;|&&|%7C%7C)cat.+(?:egory)")
```

but there may be a better way.

Also echo and cat are mislabeled as curl for the X-VSF-RuleName in cmd.vcl

**Comment by comotion** [comment]
it is greedy indeed, thanks for spotting this. to let us properly attribute this, please make a pull request. is it the only false positive you get with wordpress?

**Comment by blindpet** [comment]
This was the first false positive I've found so far, I will do more thorough testing. I have a feeling it may be easier to include the vsf.vcl after wp-admin checks (I believe most false positives will occur there). I have over 40 plugins installed so it will be a pretty thorough test once it's complete.

I've never done a PR before so I'll have to find a tutorial on it somewhere, if you have one that you can recommend I'll happily take a look. I will have to make it for both the 3.0 branch and master branch. Even though 3.0 is EOL I have a full guide for using it with Wordpress in the works.

**Comment by blindpet** [comment]
I have attempted to create pull requests https://github.com/blindpet/VSF/pull/1 and https://github.com/blindpet/VSF/pull/2

**Comment by fgsch** [comment]
@blindpet The PRs are not showing.  

I don't think the solution is correct either. For example what about "categories"?

I'd be interested to know what false positives you got since it needs to be proceeded by =, ;, && or ||. 
Can you share the details?

The other checks might be affected by this.

**Comment by blindpet** [comment]
This was in wp-admin and trying to edit the categories. I think it would be a good idea to include the vsf rules after the wordpress_logged_in check and wp-admin check

**Comment by blindpet** [comment]
I added %20 to all of the cmd.vcl and have been testing on wordpress with 40 plugins, so far everything is working, contact forms are working too. I am using the Varnish 3 branch for testing and can attempt to create a new PR.

One thing I did think of is that maybe attackers can use something other than space after shell commands but it is likely to be a special symbol, perhaps something like this is balanced

    if (req.url ~ "(=|;|&&|%7C%7C)cat%.+")

**Comment by blindpet** [comment]
There is an issue with uploading images, here is the varnishlog

` 15 TxHeader     c Connection: keep-alive
   15 TxHeader     c X-Cache: MISS
   15 Length       c 1437
   15 ReqEnd       c 1813112150 1445724958.286725760 1445724958.732088089 0.000149012 0.445240498 0.000121898
   19 Debug        c herding
   19 SessionClose c timeout
   19 StatSess     c 162.158.92.121 61863 3 1 4 0 2 2 2347 24104
   23 Debug        c herding
   23 SessionClose c timeout
   23 StatSess     c 162.158.92.121 47146 2 1 2 0 2 2 1082 1390
   15 Debug        c herding
   15 Interrupted  c ReqStart
    0 Debug        - "VCL_error(801, Rejected)"
   15 ReqStart     c 141.101.81.192 48296 1813112151
   15 RxRequest    c POST
   15 RxURL        c /wp-admin/admin-ajax.php
   15 RxProtocol   c HTTP/1.1
   15 RxHeader     c Host: www.htpcguides.com
   15 RxHeader     c Connection: Keep-Alive
   15 RxHeader     c Accept-Encoding: gzip
   15 RxHeader     c CF-IPCountry: DK
   15 RxHeader     c X-Forwarded-For: 109.198.128.118
   15 RxHeader     c CF-RAY: 23a90b1dcdc416be-ARN
   15 RxHeader     c Content-Length: 4774
   15 RxHeader     c X-Forwarded-Proto: http
   15 RxHeader     c CF-Visitor: {"scheme":"http"}
   15 RxHeader     c Accept: application/json, text/javascript, */*; q=0.01
   15 RxHeader     c Origin: http://www.htpcguides.com
   15 RxHeader     c X-Requested-With: XMLHttpRequest
   15 RxHeader     c User-Agent: Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/46.0.2490.71 Safari/537.36
   15 RxHeader     c Content-Type: application/x-www-form-urlencoded; charset=UTF-8
   15 RxHeader     c Referer: http://www.htpcguides.com/wp-admin/post.php?post=7682&action=edit
   15 RxHeader     c Accept-Language: en-US,en;q=0.8
   15 RxHeader     c Cookie: wordpress_800561c8bcb024b629356cfa3a80568e=blindpet%7C1446738937%7CkzZpWQ9eNtowa1d4t0XQGPaKaeQbPC4xi2QRwlf8SKs%7Caadff76da45796c882a4b0635fee68a2c5cc7baf187b5568a03986e7653c4aac; __cfduid=dc2e4bb8a26b89aa4c7733960f7e43f3d1444344202; ginger-cookie=
   15 RxHeader     c CF-Connecting-IP: 109.198.128.118
   15 VCL_call     c recv
   15 VCL_error    c Regexp matching returned -8
   15 VCL_return   c error
   15 VCL_call     c hash
   15 Hash         c /wp-admin/admin-ajax.php
   15 Hash         c www.htpcguides.com
   15 VCL_return   c hash
   15 VCL_call     c error
   15 VCL_Log      c vcl_error
   15 VCL_Log      c security.vcl alert xid:1813112151 HTTP/1.1 [xss-xss.xss-16][141.101.81.192] www.htpcguides.com/wp-admin/admin-ajax.php (Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/46.0.2490.71 Safari/537.36) (Cross-site Scripting)
   15 VCL_return   c deliver
   15 VCL_call     c deliver deliver
   15 TxProtocol   c HTTP/1.1
   15 TxStatus     c 403
   15 TxResponse   c Naughty, not nice!
`

**Comment by blindpet** [comment]
The wpcron error also occurs with 4.1

 << Request  >> 164258
-   Begin          req 164257 rxreq
-   Timestamp      Start: 1445733478.697552 0.000000 0.000000
-   Timestamp      Req: 1445733478.697552 0.000000 0.000000
-   ReqStart       188.166.248.150 36032
-   ReqMethod      POST
-   ReqURL         /wp-cron.php?doing_wp_cron=1445733478.6963810920715332031250
-   ReqProtocol    HTTP/1.0
-   ReqHeader      User-Agent: WordPress/4.3.1; http://188.166.248.150
-   ReqHeader      Host: 188.166.248.150
-   ReqHeader      Accept: */*
-   ReqHeader      Accept-Encoding: deflate;q=1.0, compress;q=0.5, gzip;q=0.5
-   ReqHeader      Content-Length: 0
-   ReqHeader      Content-Type: application/x-www-form-urlencoded
-   ReqHeader      X-Forwarded-For: 188.166.248.150
-   VCL_call       RECV
-   ReqHeader      X-VSF-ClientIP: 188.166.248.150
-   ReqHeader      X-VSF-Method: POST
-   ReqHeader      X-VSF-Proto: HTTP/1.0
-   ReqHeader      X-VSF-URL: 188.166.248.150/wp-cron.php?doing_wp_cron=1445733478.6963810920715332031250
-   ReqHeader      X-VSF-UA: WordPress/4.3.1; http://188.166.248.150
-   ReqUnset       X-VSF-URL: 188.166.248.150/wp-cron.php?doing_wp_cron=1445733478.6963810920715332031250
-   ReqHeader      X-VSF-URL: /wp-cron.php?doing_wp_cron=1445733478.6963810920715332031250
-   ReqHeader      X-VSF-Body:
-   ReqHeader      X-VSF-Client: [188.166.248.150] 188.166.248.150/wp-cron.php?doing_wp_cron=1445733478.6963810920715332031250 (WordPress/4.3.1; http://188.166.248.150)
-   ReqHeader      X-VSF-RuleName: Empty Content-Length Header
-   ReqHeader      X-VSF-RuleID: protocol.clen-1
-   ReqHeader      X-VSF-Rule: -protocol.clen-1
-   ReqHeader      X-VSF-Response: Naughty, not nice!
-   VCL_return     synth
-   ReqUnset       Accept-Encoding: deflate;q=1.0, compress;q=0.5, gzip;q=0.5
-   ReqHeader      Accept-Encoding: gzip
-   VCL_call       HASH
-   VCL_return     lookup
-   Timestamp      Process: 1445733478.697657 0.000105 0.000105
-   RespHeader     Date: Sun, 25 Oct 2015 00:37:58 GMT
-   RespHeader     Server: Varnish
-   RespHeader     X-Varnish: 164258
-   RespProtocol   HTTP/1.1
-   RespStatus     801
-   RespReason     Unknown HTTP Status
-   RespReason     Rejected
-   VCL_call       SYNTH
-   VCL_Log        vcl_synth
-   VCL_Log        security.vcl alert xid:164258 HTTP/1.0 [-protocol.clen-1][188.166.248.150] 188.166.248.150/wp-cron.php?doing_wp_cron=1445733478.6963810920715332031250 (WordPress/4.3.1; http://188.166.248.150) (Empty Content-Length Header)
-   RespStatus     403
-   RespReason     Forbidden
-   RespReason     Naughty, not nice!
-   RespHeader     Content-Type: text/html; charset=utf-8
-   RespHeader     Retry-After: 5
-   VCL_return     deliver
-   RespHeader     Content-Length: 277
-   Storage        malloc Transient
-   Debug          "RES_MODE 2"
-   RespHeader     Connection: close
-   Timestamp      Resp: 1445733478.697788 0.000236 0.000131
-   ReqAcct        295 0 295 204 277 481
-   End

**Comment by comotion** [comment]
Hi. the post errors, are they related to cmd.vcl? Since they aren't, I have renamed this ticket.
The 4.1 issue has to do with rule `protocol.clen-1` which should probably not trigger on missing content-length header.

**Comment by blindpet** [comment]
I should have started a new ticket, my apologies. The cmd.vcl still need to be changed for my purposes and have done so here https://github.com/blindpet/VSF/blob/3.0/vcl/rules/cmd.vcl

I spent the weekend making it wordpress compatible and the changes are in vsf.vcl https://github.com/blindpet/VSF/blob/3.0/vcl/vsf.vcl

**Comment by comotion** [comment]
yes, different issues should get their own tickets. I took a look at your changes over the weekend, and I suggest splitting site-specific rules out of vsf.vcl into their own file. Examples of this are CloudFlare and GA cookie handling, feed and user-agent as well as the WP pass rules which circumvent the VSF and any receive rules beyond this point

Regarding the protocol rules I believe I have improved upon these.

The XSS rule is more troubling though, and it would be helpful if you could capture and post the content of the POST request which matches the xss-16 rule.

**Comment by blindpet** [comment]
I did think it was a bit hacky but it got things working for 3.0 (many issues seemed to be the parseurl returning errors). Where do you suggest they get included so that appropriate security is done before their processing?

The xss-16 request no longer fails with my modifications and since it's on my production site I'd rather not do more testing. It caught something when uploading images in the wp-admin panel so the cookie check is the optimal method.

**Comment by comotion** [comment]
Oh. To test VSF in your production environment without breaking anything you can change the default handler in `handlers.vcl` to `sec_passthru`. The error will end up in varnishlog but the request will not be blocked.

**Comment by comotion** [comment]
>  Where do you suggest they get included so that appropriate security is done before their processing?

You basically have two choices, before and after the vsf include: things that have to be done before VSF (like your passthrough rules) need to be included first, and everything else should come after the vsf include.

**Comment by blindpet** [comment]
Thanks for the handler tip, that will be very useful when I do my next round of tests.

The way I have changed the vsf.vcl it basically does do the checks first and then everything else after, but I can appreciate it is better practice to make it modular. The bot checks you may want to integrate because they fail the accept empty header check, unless you can think of a better way to check for desirable  bots?

**Comment by comotion** [comment]
the rules were written to check for empy accept headers,  bots simply do not send an accept header; this has been corrected in commit 746174732d886127594b525c7ed8d765c211265a

**Comment by blindpet** [comment]
> You basically have two choices, before and after the vsf include: things that have to be done before VSF (like your passthrough rules) need to be included first, and everything else should come after the vsf include.

Adding the include anywhere besides right after vcl 4.0; at the top is fails because of the import commands. This is what I tried in the beginning of sub_recv


    if (req.url ~ "(wp-cron.php|async-upload.php|wp-content)" || (req.http.cookie ~ "wordpress_logged_in" )) {
    return (pass);
       }
    include "/etc/varnish/security/vsf.vcl";

**Comment by comotion** [comment]
You have to include vsf.vcl at the top level, outside of all sub blocks ie outside of vcl_recv {}:

```
sub vcl_recv {
  /* my vcl_recv logic */
}
include "/etc/varnish/security/vsf.vcl";
```

**Comment by blindpet** [comment]
Unfortunately that doesn't work either as I have throttle detection, so I'm guessing I have to remove import throttle from the vsf.vcl. Perhaps, in order to make things easier to move around, the user should be instructed to add import rules at the top of default.vcl?

**Comment by comotion** [comment]
The simple solution to this problem is that VSF should provide a site-specific vcl file you can edit that is included after the imports; the new commit a9624191a5e269f822e39ad683afd191277a2d84 does this.

Are we talking about the 4.0 or 3.0 branch here? Regarding "throttle detection", do you mean you use vsthrottle? Can you provide an example for how you use this? Also, what do you currently use shield for?

**Comment by comotion** [comment]
Let me rephrase my last question: do you use throttle or shield functionality on Varnish 3.0?

**Comment by blindpet** [comment]
Looks good, funnily enough after adding the include after vcl_recv all seems to be working without my additional checks in vsf.

Sorry for the confusion. I am doing 4.0 and 3.0 guides for the VSF but this specific issue is for the 3.0 branch. I use DDoS protection as outlined [here](http://www.htpcguides.com/wordpress-ddos-attack-protection-with-varnish-3-firewall/) for Varnish 3 and [here](http://www.htpcguides.com/wordpress-ddos-attack-protection-varnish-4-firewall/) for Varnish 4

**Comment by comotion** [comment]
Ah I see how that DoS protection can be useful. Note that this won't mitigate a full DDoS, but will do wonders for DoS attempts and scrapers. 

I've now added supporting code for VSF to make use of the throttle vmod in both 3.0 and master branches so that all you need to do is enable the `dos.vcl` rule.

**Comment by blindpet** [comment]
Excellent. I do get occasional false positives but have been unable to figure out why. Maybe now that it is included the static files check will change it?

One problem I saw with this is people could attack a specific static file. If that's the case consider adding some of the newer files to the static files check

    eot|ttf|cur|woff(2?)

My other concern with adding DoS protection after the standard vcl_recv is all of the processing that will be done first before throttling so I thought it was best to put at the top of vcl_recv.

Can you explain why it won't at least mitigate a full DDoS? I understand it wouldn't stop it fully but surely it would help.

**Comment by comotion** [comment]
I am actually unsure of the merit of the whitelist. Perhaps you could explain why certain URLs should not be subject to throttling? Static files, cachable by definition, should not trigger any DoS vuln in the backend anyway because delivering them does not invoke any resource cost, but the list includes stuff that I certainly wouldn't consider static at first glance.

Regarding your question about DDoS:
throttling works by storing a hit count per key, and blocking repeat offenders before they expend more resources on the backend, in this case a hit count per IP address, in memory.
 
By blocking the request before the server receives it, we also avoid an amplification attack whereby a small request can cause a large response, making it easy for a small amount of resources to create a big amount of bandwidth use.

I agree that throttling may help mitigate DDoS attacks a little, however in the event of a real DDoS attack two things happen that make this impractical:

* the amount of clients may become so great the hit counts become a prohibitively big portion of memory. This depends on the nature of the DDoS and the nature of the throttling. If you store a key per `{IP, URL}` tuple this may become a large problem fast.
* the attack often is large enough that the amount of *requests* exceeds available bandwidth to the Varnish cache, in which case mitigation strategies have to be implemented upstream of the Varnish server.

**Comment by blindpet** [comment]
When I did not have such a list of exclusions, java scripts and css files wouldn't load, fonts were hit and miss and images still trigger the calm down message some time. In addition, I had issues in wp-admin with things not loading and Firebug showed me it was because the requests were reset. The result is my workaround for things to work normally. Is there any danger to me adding a static file check before the throttle check? Something like

    if (req.url ~ “\.(jpeg|jpg|png|gif|ico|webp|woff(2?)|eot|ttf|js|css|cur|txt|pdf|gz|zip|lzma|bz2|tgz|tbz|html|htm)$”) {
    return (lookup);

Am I being paranoid about running the throttle check after the regular vcl checks consuming a lot of resources in case of an attack?

Or am I better off adding the throttle check after the rest of the vcl_recv (I use [this](http://www.htpcguides.com/configure-wordpress-varnish-3-cache-with-apache-or-nginx/) one)

I can enable the throttle check without the wp-admin etc check and provide logs if you're willing to help me diagnose this.

Thank you for the clarification regarding DDoS and how the throttle vmod works. I am behind CloudFlare too so this is just an extra step.

**Comment by comotion** [comment]
If throttling was breaking the site before whitelists, I suggest debugging the cause. Set it up to return the 429 Too Many Requests error instead of resetting the connection, and see which resources are getting 429 error code in firebug.

My guess is that the request per second value is way too restrictive - browsers will typically load far more than 15 resources from a page within 10 seconds. Style sheets and javascript usually gets hit by a single client far more than this. So a regular user hitting your site is going to get throttled!

In fact, instead of making a whitelist for everything that doesn't get throttled, I would just throttle the things that actually are a danger to DoS, like most POST endpoints, RESTful APIs and db-heavy non-cachable server generated pages. 

> Am I being paranoid about running the throttle check after the regular vcl checks consuming a lot of resources in case of an attack?

Yes. The idea is to block the requests in Varnish, and since resource consumption is so low in VCL I wouldn't worry about logic that happens before the throttle check using more resources.

A CDN like CloudFlare does far more to mitigate DDoS than you can do on one bo

**Comment by blindpet** [comment]
Thank you for all of your help with this.

Your whitelisting does make sense but I struggling to get the rule right. I've narrowed it down to search requests which use POST but I'm just checking the request URL for that. More than 1 per second seems reasonable to reset.

The issue is with admin-ajax.php requests which use POST. Generally admin-ajax.php requests need to be passed but at a reasonable rate. However, other POST requests should be limited. Can you see how to improve this?

    if(throttle.is_allowed("ip:" + req.http.X-Actual-IP, "30req/s") > 0s && (req.url ~ "\?s\=|admin-ajax.php") || (req.request == "POST"))

**Comment by comotion** [comment]
hmm. unsure. Do you mean the following?
```
  if(req.request == "POST" || req.url ~ "admin-ajax.php" || req.url ~ "wp-login.php") {
     if(throttle.is_allowed("ip:" + req.http.X-Actual-IP, "3req/s") > 0s ||
       throttle.is_allowed("ip:" + req.http.X-Actual-IP, "10req/s") > 0s) {
        error 429 "Calm down";
    }
  }
```

**Comment by comotion** [comment]
Please let us know when you write about VSF :+1:

**Comment by blindpet** [comment]
Will do, posts about software are not always well-received on github in my experience. I just want to spread your work, you are of course welcome to post the link in the wiki. Varnish 4 will be next. How should I communicate a new post about it, leave an issue on git or go in irc?

**Comment by comotion** [comment]
I appreciate posts about VSF, and also that you got in touch and therefore gave me the opportunity to improve the software. Stay tuned, more improvements are afoot (the normalization that just got commited is pretty major).

I am not much on irc these days / that is, I am but I don't read channel backlogs.
Message me on twitter (whykacper) or use github and I'll see it.

**Comment by hellracer** [comment]
Can the change be ever committed on the 3.0 branch, I think there were lot of people still on 3.x just my 0.2 cents and thanks guys for this wonderful piece of software.

**Comment by comotion** [comment]
which change are you referring to? if you mean body & normalization, then no this is not getting backported because 3.0 is missing bits to make it good enough. This is the biggest reason to phase out 3.0 !
